File: backend/app/schemas/founder.py
from datetime import date, datetime
from typing import List, Optional
from uuid import UUID

from pydantic import BaseModel, EmailStr, ConfigDict


# ---------------------------------------------------------------------------
# School management (Founder Portal: platform-wide, every school)
# ---------------------------------------------------------------------------
class SchoolManageOut(BaseModel):
    id: UUID
    name: str
    address: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None
    is_active: bool
    created_at: datetime
    total_students: int = 0
    total_teachers: int = 0
    total_classes: int = 0
    average_score: Optional[float] = None
    model_config = ConfigDict(from_attributes=True)


class SchoolCreateInput(BaseModel):
    name: str
    address: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None
    # Optionally onboard the school's Principal account in the same step.
    principal_email: Optional[EmailStr] = None
    principal_password: Optional[str] = None
    principal_full_name: Optional[str] = None


class SchoolUpdateInput(BaseModel):
    name: Optional[str] = None
    address: Optional[str] = None
    phone: Optional[str] = None
    website: Optional[str] = None


# ---------------------------------------------------------------------------
# Organization overview
# ---------------------------------------------------------------------------
class OrgOverviewOut(BaseModel):
    total_schools: int
    active_schools: int
    inactive_schools: int
    total_students: int
    total_teachers: int
    total_parents: int
    total_staff: int
    average_score: Optional[float] = None


class RoleBreakdownItem(BaseModel):
    role: str
    count: int


# ---------------------------------------------------------------------------
# Org-wide user directory (backs Founder notifications audience + user list)
# ---------------------------------------------------------------------------
class OrgUserOut(BaseModel):
    user_id: UUID
    email: str
    full_name: Optional[str] = None
    role: str
    is_active: bool
    school_id: Optional[UUID] = None
    school_name: Optional[str] = None


# ---------------------------------------------------------------------------
# Org-wide school performance (Multi-School Dashboard / School Performance)
# ---------------------------------------------------------------------------
class FounderSchoolPerformance(BaseModel):
    school_id: UUID
    school_name: str
    is_active: bool
    average_score: Optional[float] = None
    total_students: int
    total_teachers: int
    total_classes: int


# ---------------------------------------------------------------------------
# Org-wide attendance
# ---------------------------------------------------------------------------
class SchoolAttendanceSummary(BaseModel):
    school_id: UUID
    school_name: str
    present: int
    absent: int
    late: int
    total_marked: int


class OrgAttendanceOverviewOut(BaseModel):
    date: date
    present: int
    absent: int
    late: int
    total_marked: int
    by_school: List[SchoolAttendanceSummary]


# ---------------------------------------------------------------------------
# Org-wide homework
# ---------------------------------------------------------------------------
class OrgHomeworkSummary(BaseModel):
    school_id: UUID
    school_name: str
    total_assigned: int
    total_submitted: int
    total_students_covered: int
    completion_rate: Optional[float] = None


# ---------------------------------------------------------------------------
# Org-wide exams / results (one shape backs both Exam Analytics and
# Results Analytics pages, mirroring how the Principal Portal's
# exams-overview endpoint already backs two separate pages).
# ---------------------------------------------------------------------------
class OrgExamSummary(BaseModel):
    school_id: UUID
    school_name: str
    total_exams: int
    results_recorded: int
    average_score: Optional[float] = None


# ---------------------------------------------------------------------------
# Org-wide student / teacher analytics
# ---------------------------------------------------------------------------
class SchoolStudentSummary(BaseModel):
    school_id: UUID
    school_name: str
    student_count: int
    average_score: Optional[float] = None


class GradeLevelSummary(BaseModel):
    grade_level: Optional[int] = None
    count: int


class OrgStudentAnalyticsOut(BaseModel):
    by_school: List[SchoolStudentSummary]
    by_grade: List[GradeLevelSummary]


class SchoolTeacherSummary(BaseModel):
    school_id: UUID
    school_name: str
    teacher_count: int


class TopTeacher(BaseModel):
    teacher_id: UUID
    full_name: Optional[str] = None
    school_name: str
    subjects_taught: List[str] = []
    average_student_score: Optional[float] = None


class OrgTeacherAnalyticsOut(BaseModel):
    by_school: List[SchoolTeacherSummary]
    top_teachers: List[TopTeacher]

File: frontend/src/types/founder.ts
// Mirrors backend/app/schemas/founder.py

export interface SchoolManageItem {
  id: string;
  name: string;
  address: string | null;
  phone: string | null;
  website: string | null;
  is_active: boolean;
  created_at: string;
  total_students: number;
  total_teachers: number;
  total_classes: number;
  average_score: number | null;
}

export interface SchoolCreateInput {
  name: string;
  address?: string;
  phone?: string;
  website?: string;
  principal_email?: string;
  principal_password?: string;
  principal_full_name?: string;
}

export interface SchoolUpdateInput {
  name?: string;
  address?: string;
  phone?: string;
  website?: string;
}

export interface OrgOverview {
  total_schools: number;
  active_schools: number;
  inactive_schools: number;
  total_students: number;
  total_teachers: number;
  total_parents: number;
  total_staff: number;
  average_score: number | null;
}

export interface RoleBreakdownItem {
  role: string;
  count: number;
}

export interface OrgUserItem {
  user_id: string;
  email: string;
  full_name: string | null;
  role: string;
  is_active: boolean;
  school_id: string | null;
  school_name: string | null;
}

export interface FounderSchoolPerformance {
  school_id: string;
  school_name: string;
  is_active: boolean;
  average_score: number | null;
  total_students: number;
  total_teachers: number;
  total_classes: number;
}

export interface SchoolAttendanceSummary {
  school_id: string;
  school_name: string;
  present: number;
  absent: number;
  late: number;
  total_marked: number;
}

export interface OrgAttendanceOverview {
  date: string;
  present: number;
  absent: number;
  late: number;
  total_marked: number;
  by_school: SchoolAttendanceSummary[];
}

export interface OrgHomeworkSummary {
  school_id: string;
  school_name: string;
  total_assigned: number;
  total_submitted: number;
  total_students_covered: number;
  completion_rate: number | null;
}

export interface OrgExamSummary {
  school_id: string;
  school_name: string;
  total_exams: number;
  results_recorded: number;
  average_score: number | null;
}

export interface SchoolStudentSummary {
  school_id: string;
  school_name: string;
  student_count: number;
  average_score: number | null;
}

export interface GradeLevelSummary {
  grade_level: number | null;
  count: number;
}

export interface OrgStudentAnalytics {
  by_school: SchoolStudentSummary[];
  by_grade: GradeLevelSummary[];
}

export interface SchoolTeacherSummary {
  school_id: string;
  school_name: string;
  teacher_count: number;
}

export interface TopTeacher {
  teacher_id: string;
  full_name: string | null;
  school_name: string;
  subjects_taught: string[];
  average_student_score: number | null;
}

export interface OrgTeacherAnalytics {
  by_school: SchoolTeacherSummary[];
  top_teachers: TopTeacher[];
}

export interface FounderAISummary {
  data: {
    org_overview: OrgOverview;
    school_performance: FounderSchoolPerformance[];
    role_breakdown: RoleBreakdownItem[];
    attendance_today: OrgAttendanceOverview;
    risk_flags: Array<Record<string, unknown>>;
  };
  ai_recommendations: string | null;
}
